File: javascript/leetcode/653. Two Sum IV - Input is a BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function (root, k) {
  const seen = new Set();
  return has(root, k, seen);
};

function has(root, k, seen) {
  if (!root) return false;

  if (seen.has(k - root.val)) return true;

  seen.add(root.val);

  return has(root.right, k, seen) || has(root.left, k, seen);
}

File: c++/leetcode/653. Two Sum IV - Input is a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  bool has(TreeNode *, int &, unordered_set<int> &);
  bool findTarget(TreeNode *root, int k)
  {
    unordered_set<int> seen;
    return has(root, k, seen);
  }
};

bool Solution::has(TreeNode *root, int &k, unordered_set<int> &seen)
{
  if (!root)
    return false;

  if (seen.find(k - root->val) != seen.end())
    return true;

  seen.insert(root->val);
  return has(root->left, k, seen) || has(root->right, k, seen);
}